sys/dev/isp/isp.c

```diff
@@ -321,41 +383,113 @@ isp_reset(ispsoftc_t *isp, int do_load_defaults)
+	 * Early setup DMA for the request and response queues.
+	 * We do this now so we can use the request queue
+	 * for dma to load firmware from.
 	 */
-	if ((isp->isp_mdvec->dv_ispfw == NULL) || (isp->isp_confopts & ISP_CFG_NORELOAD)) {
-		dodnld = 0;
-	} else {
 
-		/*
-		 * Set up DMA for the request and response queues.
-		 * We do this now so we can use the request queue
-		 * for dma to load firmware from.
-		 */
-		if (ISP_MBOXDMASETUP(isp) != 0) {
-			isp_prt(isp, ISP_LOGERR, "Cannot setup DMA");
-			return;
+	if (ISP_MBOXDMASETUP(isp) != 0) {
+		isp_prt(isp, ISP_LOGERR, "Cannot setup DMA");
+		return;
+	}
+
+	/*
+	 * FW load priority
+	 * For 26xx and newer:
+	 * Load ispfw(4) firmware unless requested not to do so.
+	 * Request (active) flash firmware information. Compare
+	 * version numbers of ispfw(4) and flash firmware. Load
+	 * the highest version into RAM of the controller.
+	 * If loading ispfw(4) is disabled or loading it failed
+	 * (eg. no firmware available) we just load firmware from
+	 * flash. If this fails for whatever reason we fallback
+	 * to let the controller MBOX_LOAD_FLASH_FIRMWARE by itself
+	 * and hope the best to get it up and running.
+	 *
+	 * For 25xx and older:
+	 * Load ispfw(4) firmware unless requested not to do so
+	 * and load it into RAM of the controller. If loading
+	 * ispfw(4) is disabled or loading it failed (eg. no
+	 * firmware available) we just let the controller
+	 * MBOX_EXEC_FIRMWARE to start the flash firmware.
+	 */
+
+	fcparam *fcp = FCPARAM(isp, 0);
+
+	/* read FLT to get flash region addresses */
+	isp_get_flash_addrs(isp);
+
+	/* set informational sysctl(8) to sane value */
+	snprintf(fcp->fw_version_ispfw, sizeof(fcp->fw_version_ispfw),
+	    "not loaded");
+	snprintf(fcp->fw_version_flash, sizeof(fcp->fw_version_flash),
+	    "not loaded");
+	snprintf(fcp->fw_version_run, sizeof(fcp->fw_version_run),
+	    "not loaded");
+
+
+	/* Try to load ispfw(4) first */
+	if (!(isp->isp_confopts & ISP_CFG_NORELOAD)) {
+		char fwname[32];
+		snprintf(fwname, sizeof(fwname), "isp_%04x", isp->isp_did);
+		isp->isp_osinfo.ispfw = firmware_get(fwname);
+		if (isp->isp_osinfo.ispfw != NULL) {
+			isp->isp_mdvec->dv_ispfw = isp->isp_osinfo.ispfw->data;
+			const uint32_t *ispfwptr = isp->isp_mdvec->dv_ispfw;
+			for (i = 0; i < 4; i++)
+				fcp->fw_ispfwrev[i] = ispfwptr[4 + i];
+			isp_prt(isp, ISP_LOGCONFIG,
+			    "Loaded ispfw(4) firmware %s", fwname);
+			snprintf(fcp->fw_version_ispfw,
+			    sizeof(fcp->fw_version_ispfw),
+			    "%u.%u.%u", fcp->fw_ispfwrev[0],
+			    fcp->fw_ispfwrev[1], fcp->fw_ispfwrev[2]);
+			isp_prt(isp, ISP_LOGCONFIG,
+			    "Firmware revision (ispfw) %u.%u.%u (%x).",
+			    fcp->fw_ispfwrev[0], fcp->fw_ispfwrev[1],
+			    fcp->fw_ispfwrev[2], fcp->fw_ispfwrev[3]);
+		} else {
+			snprintf(fcp->fw_version_ispfw,
+			    sizeof(fcp->fw_version_ispfw), "load error");
+			isp_prt(isp, ISP_LOGERR,
+			    "Error loading ispfw(4) firmware %s", fwname);
```

Comment:
It may be either due to firmware absence or ```firmware_get()``` failed for whatever reason.
As ```firmware_get()``` is unspecific and just returns ```NULL``` on any failure we can not further distinguish.
Could be changed to ```ISP_LOGWARN``` if this fits better in your eyes.

Reply:
It would not change anything, since both WARN and ERR are enabled by default.  I'd prefer to to be rephrased somehow to not look like an error, since I guess in most cases it will just mean that ispfw(4) is not loaded, which is not an error.

Reply:
> t would not change anything, since both WARN and ERR are enabled by default.

You're right. So
```
			isp_prt(isp, ISP_LOGWARN,
			    "Unable to load ispfw(4) firmware %s", fwname);
```
or better hide this behind `LOGDEBUG0`?
As we don't talk about an error anymore, those lines above
```
			snprintf(fcp->fw_version_ispfw,
			    sizeof(fcp->fw_version_ispfw), "load error");
```
can be removed to, as `fcp->fw_version_ispfw` is already set to `not loaded` which is still valid.